### ISSUE TITLE
aws_security_group_rule Using multiple times the same port range and protocol causes problems with the state

### ISSUE THREAD

**Issue by robertfirek** [hashicorp/terraform#3066]
Security group rules are not saved in the state when ***aws_security_group_rules*** is applied multiple times for the same port range and protocol. The rules are created properly in Amazon, but the state doesn't contains any entries. 

When rules is merged to one rules the state is updated properly.

Probably there should not be allowed to create rules in this way or implementation should handle such a case. 

Terraform version: 0.6.3

***Example:***
Defining security role for port ***9999*** and ***tcp*** creates rules in Amazon, but the state doesn't contain any information about it. 

```
resource "aws_security_group_rule" "ingress_no_1" {
    type        = "ingress"
    from_port   = 9999
    to_port     = 9999
    protocol    = "tcp"
    cidr_blocks = ["${var.first_cidr_block}"]

    security_group_id = "${aws_security_group.my_security_group.id}"
}

resource "aws_security_group_rule" "ingress_no_2" {
    type        = "ingress"
    from_port   = 9999
    to_port     = 9999
    protocol    = "tcp"
    cidr_blocks = ["${var.second_cidr_block}"]

    security_group_id = "${aws_security_group.my_security_group.id}"
}

resource "aws_security_group_rule" "ingress_no_3" {
    type        = "ingress"
    from_port   = 9999
    to_port     = 9999
    protocol    = "tcp"
    cidr_blocks = ["${var.third_cidr_block}"]

    security_group_id = "${aws_security_group.my_security_group.id}"
}
```

Merged version works without any issues.

```
resource "aws_security_group_rule" "ingress_no_1" {
    type        = "ingress"
    from_port   = 9999
    to_port     = 9999
    protocol    = "tcp"
    cidr_blocks = ["${var.first_cidr_block}", "${var.second_cidr_block}", "${var.third_cidr_block}"]

    security_group_id = "${aws_security_group.my_security_group.id}"
}
```

**Comment by pmvilaca** [comment]
I also confirm the bug when we're using the aws_security_group and adding a list of ingress objects with the same port.

```
resource "aws_security_group" "office_ssh" {
  name = "office_sg"
  description = "Enable SSH connections from our offices"
  vpc_id = "${aws_vpc.main.id}"

  ingress {
    from_port = 22
    to_port = 22
    protocol = "tcp"
    cidr_blocks = [
      "x.x.x.x/32"
    ]
  }
  ingress {
    from_port = 22
    to_port = 22
    protocol = "tcp"
    cidr_blocks = [
      "y.y.y.y32"]
  }
  ingress {
    from_port = 22
    to_port = 22
    protocol = "tcp"
    cidr_blocks = [
      "z.z.z.z/32"]
  }
}
```

The workaround and btw, the correct way to represent this is:

```
resource "aws_security_group" "office_ssh" {
  name = "office_sg"
  description = "Enable SSH connections from our offices"
  vpc_id = "${aws_vpc.main.id}"

  ingress {
    from_port = 22
    to_port = 22
    protocol = "tcp"
    cidr_blocks = [
      "x.x.x.x/32",
      "y.y.y.y/32",
      "z.z.z.z/32"
    ]
  }
}
```

**Comment by catsby** [comment]
Hey Friends – sorry for the very long silence here! I'm happy to say that this (these?) was recently patched in https://github.com/hashicorp/terraform/pull/5100, and went out in v0.6.12 (latest version at this time being v0.6.15). Sorry for the long wait! 

Let us know if you need anything else!

Best,
Clint

**Comment by chunnu** [comment]
Is this issue fixed? I am using terraform version 0.6.16 but still getting this error:

**aws_security_group.cassandra: Error authorizing security group ingress rules: InvalidParameterValue: The same permission must not appear multiple times
	status code: 400, request id: afd8e9d6-71f3-4ad7-a3c3-58ca41f29945**

I am using the same port and protocol but different CIDRS in 2 separate SG resource and hitting the above error.

The above comment says thats its fixed in 0.6.12. But I don't find that in the terraform CHANGELOG.

Can someone point me to the correct version in which it is fixed.

**Comment by abhishek063** [comment]
Is this issue fixed? I am using terraform version 0.6.16 but still getting this error:

aws_security_group.cassandra: Error authorizing security group ingress rules: InvalidParameterValue: The same permission must not appear multiple times
status code: 400, request id: afd8e9d6-71f3-4ad7-a3c3-58ca41f29945

I am using the same port and protocol but different CIDRS in 2 separate SG resource and hitting the above error.

The above comment says thats its fixed in 0.6.12. But I don't find that in the terraform CHANGELOG.

Can someone point me to the correct version in which it is fixed.
